File: backup/DirectDeposit.js
import React, { useState, useEffect } from 'react';

const BASE = process.env.REACT_APP_API_URL || 'http://localhost:8000';
const tkn = () => localStorage.getItem('payroll_token');
async function req(path, opts = {}) {
  const res = await fetch(`${BASE}${path}`, {
    ...opts, headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${tkn()}`, ...opts.headers },
  });
  if (!res.ok) { const e = await res.json().catch(() => ({ detail: res.statusText })); throw new Error(e.detail); }
  if (res.status === 204) return null;
  return res.json();
}

export default function DirectDeposit() {
  const [summary, setSummary] = useState(null);
  const [employees, setEmployees] = useState([]);
  const [selected, setSelected] = useState('');
  const [bankInfo, setBankInfo] = useState(null);
  const [form, setForm] = useState({ routing_number: '', account_number: '', account_type: 'checking', bank_name: '' });
  const [showForm, setShowForm] = useState(false);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState('');
  const [success, setSuccess] = useState('');

  useEffect(() => {
    Promise.all([req('/direct-deposit/summary'), req('/employees?status=active')]).then(([s, e]) => {
      setSummary(s); setEmployees(e?.employees || []);
    });
  }, []);

  const loadBank = async (id) => {
    if (!id) { setBankInfo(null); return; }
    try { setBankInfo(await req(`/direct-deposit/employees/${id}`)); }
    catch { setBankInfo(null); }
  };

  const save = async (e) => {
    e.preventDefault(); setError(''); setSuccess('');
    setLoading(true);
    try {
      await req(`/direct-deposit/employees/${selected}`, { method: 'POST', body: JSON.stringify(form) });
      setSuccess('Bank account saved'); setShowForm(false);
      setForm({ routing_number: '', account_number: '', account_type: 'checking', bank_name: '' });
      loadBank(selected);
    } catch (err) { setError(err.message); }
    setLoading(false);
  };

  const verify = async () => {
    try { await req(`/direct-deposit/employees/${selected}/verify`, { method: 'PUT' }); loadBank(selected); setSuccess('Verified'); }
    catch (err) { setError(err.message); }
  };

  const remove = async () => {
    if (!window.confirm('Remove bank account?')) return;
    try { await req(`/direct-deposit/employees/${selected}`, { method: 'DELETE' }); setBankInfo(null); setSuccess('Removed'); }
    catch (err) { setError(err.message); }
  };

  const enrollPct = summary ? Math.round((summary.enrolled / Math.max(summary.total_employees, 1)) * 100) : 0;

  return (
    <div className="page">
      <div className="page-header"><h1>Direct Deposit</h1></div>
      {error && <div className="alert alert-danger">{error}</div>}
      {success && <div className="alert" style={{ background: 'var(--green-bg)', color: 'var(--green)', border: '1px solid #b2dfb2', padding: '10px 14px', borderRadius: 6, marginBottom: 12, fontSize: 13 }}>✓ {success}</div>}

      {/* Summary */}
      {summary && (
        <div className="metrics-grid" style={{ gridTemplateColumns: 'repeat(4,1fr)', marginBottom: 16 }}>
          {[
            ['Total employees', summary.total_employees],
            ['Enrolled', summary.enrolled],
            ['Verified', summary.verified],
            ['Not enrolled', summary.not_enrolled],
          ].map(([label, val]) => (
            <div key={label} className="metric-card">
              <div className="metric-label">{label}</div>
              <div className="metric-value">{val}</div>
            </div>
          ))}
        </div>
      )}
      {summary && (
        <div className="card" style={{ marginBottom: 16 }}>
          <div style={{ padding: 16 }}>
            <div style={{ display: 'flex', justifyContent: 'space-between', fontSize: 12, marginBottom: 6 }}>
              <span style={{ color: 'var(--text2)' }}>Enrollment rate</span>
              <span style={{ fontWeight: 500 }}>{enrollPct}%</span>
            </div>
            <div style={{ height: 8, background: 'var(--bg3)', borderRadius: 99, overflow: 'hidden' }}>
              <div style={{ height: '100%', width: `${enrollPct}%`, background: enrollPct === 100 ? 'var(--green)' : 'var(--blue)', borderRadius: 99, transition: 'width 0.4s' }} />
            </div>
          </div>
        </div>
      )}

      <div className="two-col">
        {/* Employee selector + form */}
        <div className="card">
          <div className="card-header"><h3>Manage bank account</h3></div>
          <div style={{ padding: 16 }}>
            <div className="form-group">
              <label>Employee</label>
              <select value={selected} onChange={e => { setSelected(e.target.value); loadBank(e.target.value); setBankInfo(null); setShowForm(false); }}>
                <option value="">Select employee…</option>
                {employees.map(e => <option key={e.id} value={e.id}>{e.full_name}</option>)}
              </select>
            </div>

            {selected && bankInfo && !bankInfo.has_direct_deposit && (
              <div style={{ marginTop: 12 }}>
                <p style={{ fontSize: 13, color: 'var(--text3)', marginBottom: 12 }}>No bank account on file</p>
                <button className="btn btn-primary" onClick={() => setShowForm(true)}>+ Add bank account</button>
              </div>
            )}

            {selected && bankInfo?.has_direct_deposit && !showForm && (
              <div style={{ marginTop: 12 }}>
                <div className="preview-box" style={{ marginBottom: 12 }}>
                  {[
                    ['Bank', bankInfo.bank_name || '—'],
                    ['Account type', bankInfo.account_type],
                    ['Account number', bankInfo.account_display],
                    ['Routing number', bankInfo.routing_display],
                    ['Status', bankInfo.is_verified ? '✓ Verified' : '⚠ Pending verification'],
                  ].map(([l, v]) => (
                    <div key={l} className="preview-row">
                      <span style={{ color: 'var(--text2)' }}>{l}</span>
                      <span style={{ color: v.startsWith('✓') ? 'var(--green)' : v.startsWith('⚠') ? 'var(--amber)' : 'inherit' }}>{v}</span>
                    </div>
                  ))}
                </div>
                <div style={{ display: 'flex', gap: 8 }}>
                  {!bankInfo.is_verified && <button className="btn" onClick={verify}>✓ Mark verified</button>}
                  <button className="btn" onClick={() => setShowForm(true)}>Update account</button>
                  <button className="btn btn-danger" onClick={remove}>Remove</button>
                </div>
              </div>
            )}

            {showForm && (
              <form onSubmit={save} style={{ marginTop: 16 }}>
                <div className="form-group"><label>Bank name</label><input type="text" value={form.bank_name} onChange={e => setForm(f => ({ ...f, bank_name: e.target.value }))} placeholder="Chase, Wells Fargo, etc." /></div>
                <div className="form-group">
                  <label>Account type</label>
                  <select value={form.account_type} onChange={e => setForm(f => ({ ...f, account_type: e.target.value }))}>
                    <option value="checking">Checking</option>
                    <option value="savings">Savings</option>
                  </select>
                </div>
                <div className="form-group"><label>Routing number (9 digits) *</label><input type="text" value={form.routing_number} onChange={e => setForm(f => ({ ...f, routing_number: e.target.value }))} placeholder="021000021" maxLength={9} required /></div>
                <div className="form-group"><label>Account number *</label><input type="text" value={form.account_number} onChange={e => setForm(f => ({ ...f, account_number: e.target.value }))} placeholder="4-17 digits" required /></div>
                <div style={{ display: 'flex', gap: 8, marginTop: 8 }}>
                  <button className="btn btn-primary" type="submit" disabled={loading}>{loading ? 'Saving…' : 'Save bank account'}</button>
                  <button className="btn" type="button" onClick={() => setShowForm(false)}>Cancel</button>
                </div>
              </form>
            )}
          </div>
        </div>

        {/* Instructions */}
        <div className="card">
          <div className="card-header"><h3>About direct deposit</h3></div>
          <div style={{ padding: 16, fontSize: 13, color: 'var(--text2)', lineHeight: 1.8 }}>
            <div style={{ fontWeight: 500, marginBottom: 8 }}>How it works:</div>
            <div>1. Collect employee bank account info here</div>
            <div>2. Verify routing number via ABA checksum</div>
            <div>3. Send micro-deposits for verification (2–3 business days)</div>
            <div>4. Employee confirms micro-deposit amounts</div>
            <div>5. Account marked verified — ready for ACH payroll</div>
            <div style={{ marginTop: 16, padding: 12, background: 'var(--amber-bg)', borderRadius: 6, fontSize: 12, color: 'var(--amber)' }}>
              ⚠ Actual ACH transfers require a banking partner (Dwolla, Stripe Payouts, or payroll processor). This system securely stores account info for when you integrate one.
            </div>
            <div style={{ marginTop: 12, fontSize: 12, color: 'var(--text3)' }}>
              Bank routing and account numbers are encrypted at rest using AES-256-GCM. Set SSN_ENCRYPTION_KEY in .env to enable encryption.
            </div>
          </div>
        </div>
      </div>
    </div>
  );
}

File: backup/direct_deposit.py
"""
Direct deposit bank account management.
Stores employee bank account info for payroll processing.
Actual ACH transfers require a banking partner (Dwolla, Stripe Payouts, etc.)

Routing + account numbers are AES-encrypted at rest using the same
encryption service as SSNs.

POST   /direct-deposit/employees/{id}   add/update bank account
GET    /direct-deposit/employees/{id}   get masked bank info
DELETE /direct-deposit/employees/{id}   remove bank account
GET    /direct-deposit/employees/{id}/verify   check account is verified
"""
import uuid
from datetime import datetime
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import Column, String, Boolean, DateTime, ForeignKey, select
from sqlalchemy.dialects.postgresql import UUID
from pydantic import BaseModel
from database import Base, get_db
from services.encryption import encrypt_ssn as _encrypt, decrypt_ssn as _decrypt
from utils.auth import get_current_user

router = APIRouter(prefix="/direct-deposit", tags=["direct-deposit"])


class BankAccount(Base):
    __tablename__ = "bank_accounts"
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    employee_id = Column(UUID(as_uuid=True), ForeignKey("employees.id", ondelete="CASCADE"), unique=True)
    company_id = Column(UUID(as_uuid=True), ForeignKey("companies.id", ondelete="CASCADE"))
    bank_name = Column(String(100))
    account_type = Column(String(20), default="checking")   # checking | savings
    routing_number_encrypted = Column(String(500))          # AES-encrypted
    account_number_encrypted = Column(String(500))          # AES-encrypted
    account_last4 = Column(String(4))                       # stored unencrypted for display
    routing_last4 = Column(String(4))
    is_verified = Column(Boolean, default=False)
    is_active = Column(Boolean, default=True)
    added_at = Column(DateTime(timezone=True), default=datetime.utcnow)
    verified_at = Column(DateTime(timezone=True))


class BankAccountCreate(BaseModel):
    bank_name: Optional[str] = None
    account_type: str = "checking"
    routing_number: str
    account_number: str


@router.get("/employees/{employee_id}")
async def get_bank_account(
    employee_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    result = await db.execute(
        select(BankAccount).where(
            BankAccount.employee_id == employee_id,
            BankAccount.company_id == current_user["company_id"],
            BankAccount.is_active == True,
        )
    )
    acct = result.scalar_one_or_none()
    if not acct:
        return {"has_direct_deposit": False}
    return {
        "has_direct_deposit": True,
        "bank_name": acct.bank_name,
        "account_type": acct.account_type,
        "routing_last4": acct.routing_last4,
        "account_last4": acct.account_last4,
        "account_display": f"****{acct.account_last4}",
        "routing_display": f"****{acct.routing_last4}",
        "is_verified": acct.is_verified,
        "added_at": str(acct.added_at),
    }


@router.post("/employees/{employee_id}", status_code=201)
async def add_bank_account(
    employee_id: str,
    body: BankAccountCreate,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    # Validate routing number (9 digits, ABA checksum)
    routing = body.routing_number.strip().replace("-", "")
    if not routing.isdigit() or len(routing) != 9:
        raise HTTPException(400, "Routing number must be 9 digits")
    if not _aba_checksum(routing):
        raise HTTPException(400, "Invalid routing number (ABA checksum failed)")

    account = body.account_number.strip().replace("-", "").replace(" ", "")
    if not account.isdigit() or not (4 <= len(account) <= 17):
        raise HTTPException(400, "Account number must be 4-17 digits")

    # Deactivate existing
    existing_res = await db.execute(
        select(BankAccount).where(
            BankAccount.employee_id == employee_id,
            BankAccount.company_id == current_user["company_id"],
        )
    )
    for existing in existing_res.scalars().all():
        existing.is_active = False

    # Store encrypted
    acct = BankAccount(
        employee_id=employee_id,
        company_id=current_user["company_id"],
        bank_name=body.bank_name,
        account_type=body.account_type,
        routing_number_encrypted=_safe_encrypt(routing),
        account_number_encrypted=_safe_encrypt(account),
        account_last4=account[-4:],
        routing_last4=routing[-4:],
        is_verified=False,
    )
    db.add(acct)
    await db.commit()
    await db.refresh(acct)

    return {
        "message": "Bank account saved",
        "account_display": f"****{acct.account_last4}",
        "routing_display": f"****{acct.routing_last4}",
        "account_type": acct.account_type,
        "is_verified": False,
        "note": "Bank account stored securely. Micro-deposit verification will be sent.",
    }


@router.put("/employees/{employee_id}/verify")
async def verify_bank_account(
    employee_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    """Mark bank account as verified (in production: verify micro-deposits)."""
    result = await db.execute(
        select(BankAccount).where(
            BankAccount.employee_id == employee_id,
            BankAccount.company_id == current_user["company_id"],
            BankAccount.is_active == True,
        )
    )
    acct = result.scalar_one_or_none()
    if not acct:
        raise HTTPException(404, "No bank account found")
    acct.is_verified = True
    acct.verified_at = datetime.utcnow()
    await db.commit()
    return {"message": "Bank account verified", "is_verified": True}


@router.delete("/employees/{employee_id}", status_code=204)
async def remove_bank_account(
    employee_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    result = await db.execute(
        select(BankAccount).where(
            BankAccount.employee_id == employee_id,
            BankAccount.company_id == current_user["company_id"],
        )
    )
    for acct in result.scalars().all():
        acct.is_active = False
    await db.commit()


@router.get("/summary")
async def direct_deposit_summary(
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    """Summary of direct deposit enrollment across all employees."""
    from models import Employee
    from sqlalchemy import func

    emp_count_res = await db.execute(
        select(func.count(Employee.id)).where(
            Employee.company_id == current_user["company_id"],
            Employee.status == "active",
        )
    )
    total_employees = emp_count_res.scalar() or 0

    enrolled_res = await db.execute(
        select(func.count(BankAccount.id)).where(
            BankAccount.company_id == current_user["company_id"],
            BankAccount.is_active == True,
        )
    )
    enrolled = enrolled_res.scalar() or 0

    verified_res = await db.execute(
        select(func.count(BankAccount.id)).where(
            BankAccount.company_id == current_user["company_id"],
            BankAccount.is_active == True,
            BankAccount.is_verified == True,
        )
    )
    verified = verified_res.scalar() or 0

    return {
        "total_employees": total_employees,
        "enrolled": enrolled,
        "verified": verified,
        "not_enrolled": total_employees - enrolled,
        "enrollment_pct": round(enrolled / total_employees * 100, 1) if total_employees else 0,
    }


def _aba_checksum(routing: str) -> bool:
    """Validate 9-digit ABA routing number via checksum."""
    try:
        d = [int(c) for c in routing]
        checksum = (
            3 * (d[0] + d[3] + d[6]) +
            7 * (d[1] + d[4] + d[7]) +
            1 * (d[2] + d[5] + d[8])
        )
        return checksum % 10 == 0
    except Exception:
        return False


def _safe_encrypt(value: str) -> str:
    """Encrypt using services.encryption, fall back to storing raw if key not set."""
    try:
        return _encrypt(value)
    except Exception:
        return f"plain:{value}"
